concept/thing/attribute.rs

```diff
@@ -119,35 +119,43 @@ impl<'a> ThingAPI<'a> for Attribute<'a> {
         Attribute { vertex, value: None }
     }
 
-    fn vertex(&self) -> AttributeVertex<'_> {
+    fn vertex(&self) -> Self::Vertex<'_> {
         self.vertex.as_reference()
     }
 
-    fn into_vertex(self) -> AttributeVertex<'a> {
+    fn into_vertex(self) -> Self::Vertex<'a> {
         self.vertex
     }
 
     fn into_owned(self) -> Self::Owned {
         Attribute::new(self.vertex.into_owned())
     }
 
-    fn set_modified(&self, snapshot: &mut impl WritableSnapshot, thing_manager: &ThingManager) {
-        debug_assert_eq!(thing_manager.get_status(snapshot, self.vertex().as_storage_key()), ConceptStatus::Put);
-        // Attributes are always PUT, so we don't have to record a lock on modification
+    fn set_required(&self, snapshot: &mut impl WritableSnapshot, thing_manager: &ThingManager) {
+        match self.type_().get_value_type(snapshot, thing_manager.type_manager()).ok() {
+            Some(Some(value_type)) => match value_type {
+                | ValueType::Boolean
+                | ValueType::Long
+                | ValueType::Double
+                | ValueType::Decimal
+                | ValueType::Date
+                | ValueType::DateTime
+                | ValueType::DateTimeTZ
+                | ValueType::Duration => snapshot.put(self.vertex().as_storage_key().into_owned_array()),
+                // ValueTypes with expensive writes
+                | ValueType::String | ValueType::Struct(_) => {
+                    thing_manager.lock_existing_attribute(snapshot, self.as_reference())
+                }
+            },
+            // Read error or empty value type
+            _ => snapshot.put(self.vertex().as_storage_key().into_owned_array()),
```

Comment:
This should be an impossible state right? Impossible/unexpected states we should as a policy make the system crash - we don't want it to move along in an invalid state. This will force us to address it quickly as well.

i would just say `panic!("Attribute instances must have a value type.)`.

However, the ConceptReadError you can actually propagate up to the caller since that could be something like the database is deleted/storage shutdown/OS error (we haven't quite decided how to handle these yet, so we're sticking to propagating them back)

Reply:
Good point, fixed. Hadn't wanted to return Result from this method for some reason.